### PR TITLE
Feature(Issue #5) : Join an existing shopping list to allow list collaboration

### LINKED ISSUE #5: 5. As a user, I want to join an existing shopping list so I can share a shopping list with another person.
Tokens can be shared with other users to allow them to co-manage a given list.

Joining an existing list consists of the following:

* User enters an existing token
* Save the token to `localStorage`
* Show the user the existing list

AC:

* When the user does not already have a token in `localStorage`, on the onboarding/home screen, a simple form is displayed that allows the user to enter a token
* Entering the token and hitting submit saves the token to `localStorage`, effectively giving them joint control of the list
* On submit, show an error if the token does not exist
* If they get an error message, allow them to try again or create a new list

### PR DESCRIPTION
**Pull request by daredtech**
## Description

This pull request allows the user to join the existing shopping list by providing the token associated with this list. If there is no such list in the database, an error would be displayed, allowing the user to re-enter the token or create a new list. Moreover, previously accessed lists would be displayed so the user could pick which one they want to modify. 

- Retired the `List` and `Add Item` links from the `Footer` component. 
- Added `Lists` component that would store and render all the available tokens. Clicking on a token would take you to the list associated with it, display the items in a current list, and allow the user to add new items. 
- Added `Footer` component to the `List` component and `Add Item` component that allows the user to go on the main page and change the active list. When the user clicks `Create a new shopping list` or `Join an existing list` buttons, the token gets saved to the local storage with the corresponding key.  

## Related Issue

Closes #5

## Acceptance Criteria

- [X]  When the user does not already have a token in localStorage, on the onboarding/home screen, a simple form is displayed that allows the user to enter a token
- [X]  Entering the token and hitting submit saves the token to localStorage, effectively giving them joint control of the list
- [X]  On submit, show an error if the token does not exist
- [X]  If they get an error message, allow them to try again or create a new list

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|     | :bug: Bug fix              |
| ✓  | :sparkles: New feature     |
|  ✓ | :hammer: Refactoring       |
|     | :100: Add tests            |
|     | :link: Update dependencies |
|     | :scroll: Docs              |

### Before

image

### After

image

## Testing Steps / QA Criteria

- From your terminal, run `git checkout bl-vh-join-and-share-shopping-list`.
- Once on the `bl-vh-join-and-share-shopping-list` branch, run `npm start` to load the browser.
- Enter the existing token `andes casey mew` and click `Join an existing list` button to join the list. Select the token name from `Available Lists` and click on it. This would take you to the chosen list and display the items in it. Once done adding items, click `Home` to select a different list or create a new one.
- Enter non-existing token `i don't exist` and click `Join an existing list` button to join the list. An alert would be displayed saying `Sorry, this collection doesn't exist. Please try again or create a new list`.

### CODE REVIEW

File: src/components/CustomButton/CustomButton.component.jsx [hunk 1]
```diff
@@ -3,7 +3,7 @@ import React from 'react';
 import './CustomButton.style.scss';
 
 // Allows us to have an easy to use button element and pass props for custom styling later on.
-const CustomButton = ({ children }) => (
+const CustomButton = ({ children, ...otherProps }) => (
```
Comment (jendevelops) [inline src/components/CustomButton/CustomButton.component.jsx:6]: Noice use of destructuring 👍👍👍👍👍👍👍👍

File: src/components/Home/Home.component.jsx [hunk 2]
```diff
@@ -1,46 +1,77 @@
...
-  const generateToken = () => {
-    console.log('called generate token');
+  const joinExistingList = async () => {
+    let tokenRef = firebase.dataBase.collection(`${tokenValue}`);
+    let snapshot = await tokenRef.get();
 
+    if (snapshot.empty) {
+      alert(
+        "Sorry, this collection doesn't exist. Please try again or create a new list.",
+      );
```
Comment (jendevelops) [inline src/components/Home/Home.component.jsx:22]: Javascript alert messages are not the most accessible option when displaying an error.

https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/forms/alerts#Checking_for_validity_and_notifying_the_user_2

Maybe displaying errors is something we can discuss during the Weekly Sync about how to approach it in a different way. @joeylaguna

File: src/components/Home/Home.component.jsx [hunk 3]
```diff
@@ -1,46 +1,77 @@
...
+      </p>
+      <FormInput onChange={onChange} label={'Share Token'} value={tokenValue} />
+      <button
+        onClick={joinExistingList}
+        className={`${!tokenValue ? 'button--disabled' : ''} page_button`}
+      >
+        {`
+        ${
+          !tokenValue ? 'Please enter List name first' : 'Join an existing list'
+        }`}
```
Comment (jendevelops) [inline src/components/Home/Home.component.jsx]: Love the use of template literals scattered throughout your code! Efficient!

File: src/components/Home/Home.component.jsx [hunk 4]
```diff
@@ -1,46 +1,77 @@
...
-  const redirectAddItem = () => {
-    history.push('/addItem');
+
+  //To add the token to the storage
+  const addTokenToStorage = token => {
+    //To set the item to the local storage
+    localStorage.setItem(token, token);
+  };
+
+  // To
```
Comment (joeylaguna) [inline src/components/Home/Home.component.jsx]: Can we add something to this comment or remove it?

Comment (jendevelops) [reply src/components/Home/Home.component.jsx]: Beat me to it! 😅

File: src/components/Lists/Lists.style.scss [hunk 5]
```diff
@@ -0,0 +1,23 @@
+.lists {
```
Comment (joeylaguna) [inline src/components/Lists/Lists.style.scss:1]: Always been a fan of having a separate CSS file 😄

File: src/components/Home/Home.component.jsx [hunk 6]
```diff
@@ -1,46 +1,77 @@
...
 const Home = () => {
-  const [localToken, setLocalToken] = useState(localStorage.getItem('token'));
+  const [localToken, setLocalToken] = useState('');
+  const [tokenValue, setTokenValue] = useState('');
+
   const history = useHistory();
 
-  const generateToken = () => {
-    console.log('called generate token');
+  const joinExistingList = async () => {
```
Comment (joeylaguna) [inline src/components/Home/Home.component.jsx:15]: Great use of async/await here!

File: src/components/Home/Home.component.jsx [hunk 7]
```diff
@@ -1,46 +1,77 @@
...
+
 const Home = () => {
-  const [localToken, setLocalToken] = useState(localStorage.getItem('token'));
+  const [localToken, setLocalToken] = useState('');
+  const [tokenValue, setTokenValue] = useState('');
+
   const history = useHistory();
 
-  const generateToken = () => {
-    console.log('called generate token');
```
Comment (jendevelops) [inline src/components/Home/Home.component.jsx:9]: Is this console log necessary?

**Review by jendevelops — CHANGES_REQUESTED** [review]
I pulled down the branch and attempted to connect to a share token that's on firebase, but I was not redirected to the shopping list. @daredtech @brewswain Do you have any idea what the issue could be?

Another thing I noticed is that clicking join an existing list generates a new token locally, but I don't see a new collection being generated on the Firebase side.

Screenshot from 2020-08-22 17-46-04

You can see all the tokens I generated locally, but they're not present in Firebase

Screenshot from 2020-08-22 17-47-38

**Comment by daredtech** [comment]
Good catch, @jendevelops ! Seems like we have an issue with Available Lists in `Lists` component. Right now, it displays all the keys available in one's localStorage. However, instead, it should check them against the database and only display the ones that exist in the database. We are going to fix that. 

`Join an existing list` is not intended to generate a new collection in the database. On the contrary, it checks whether there is a collection associated with the token entered and only then saves it your local storage so one would have access to it. Only `Create a New Shopping List` generates a new token and updates the collection in the database

File: src/components/Home/Home.component.jsx (hunk 4 above)
Comment (brewswain) [reply src/components/Home/Home.component.jsx]: Oops 😳
I'll send a commit now

File: src/components/Home/Home.component.jsx (hunk 7 above)
Comment (brewswain) [reply src/components/Home/Home.component.jsx:9]: Oh yeah we removed it cause we didn't think it was needed!

**Comment by jendevelops** [comment]
Thanks for the reply @daredtech @brewswain 

I also had issues when creating a new token. It would generate a new token locally, but I would not see a new token in firebase. Are y'all able to recreate that issue?

**Comment by CodeMeKathy** [comment]
Overall, great contributions and styling @brewswain and @daredtech!

@jendevelops, when I create a shopping list (token: berg aztec jew) but do not add an item to the list, I experience something similar. 

image

image


> Thanks for the reply @daredtech @brewswain
> 
> I also had issues when creating a new token. It would generate a new token locally, but I would not see a new token in firebase. Are y'all able to recreate that issue?

**Review by CodeMeKathy — APPROVED** [review]

**Comment by jendevelops** [comment]
@CodeMeKathy Ah, I see that is the issue. If that's the case that makes sense. Don't want to clutter up firbase too much with empty lists.

**Comment by daredtech** [comment]
@CodeMeKathy @jendevelops That is an inherited functionality. Firestore gets updated only when there is at least one item added to the associated list/token. Otherwise, it could get populated with a bunch of empty lists. We could discuss it more during our call today and come up with an ultimate implementation.

**Comment by jendevelops** [comment]
@brewswain @daredtech After the weekly sync today, I have a much better idea of what you were trying to go for in terms of functionality. A couple of points I wanted to bring up:

1. If we let the user input a token and it gets stored in local storage, but the list is synced with Firebase until an item is added, what will happen if someone ends up typing in the same token, and starts using the list? Or what happens if person 2 generates a new list that has the same token, and down the line person 1 tries to add items to that list as well?

2. Regarding usability: it was not entirely clear to me that joining a list would not automatically take me to that list. It would be nice if joining a list either:
    - adds to the available lists and there's a button that says "go to list"
    - redirects you straight to the list

Other than those two points, really great job with the idea to add the Available Lists section to the home page. Love it!

**Review by stef-codes — CHANGES_REQUESTED** [review]
- I was unable to see the lists. I tried with an incognito browser and with my existing browser that has an existing token. 
- They key and the values seem to be mixed in local storage (This could be a mismatch from the implementation in another issue.)

This is an example with multiple tokens: 

Screen Shot 2020-08-23 at 5 41 06 PM

**Comment by brewswain** [comment]
> * I was unable to see the lists. I tried with an incognito browser and with my existing browser that has an existing token.
> * They key and the values seem to be mixed in local storage (This could be a mismatch from the implementation in another issue.)
> 
> This is an example with multiple tokens:
> 
> Screen Shot 2020-08-23 at 5 41 06 PM

Hey stefanie, this is cause there's a bit of a restructuring going on, so you def won't see any lists right now :(

**Comment by daredtech** [comment]
@jendevelops 

1. At the current state, everyone with the same token would have access to the same list. Let's assume the initial list had 'coffee, creamer' when user1 opened it. User1 wants to add 'oat milk' and gets distracted. Meanwhile, user2 adds 'ice' to the list. The moment the firebase gets updated with 'ice', the list would re-render to display the change for everyone who uses the same token.  The updated list would be 'coffee, creamer, ice'. In the future, we could add a password, so only authorized people could share the list. Moreover, we could ensure that all the tokens are unique. But it is outside of this issue.
2. We could change the button to `Add token` and add a description something like 'click on the token associated with a list you want to edit' or something. Would that be acceptable?
Thanks.

**Comment by daredtech** [comment]
Fixed token validation. 
Only tokens that exist in the database would get displayed now. 
Added a note about clicking on a token.

**Pull request merged**